## Joel Brigida

## Assignment 6b

## Polynomial Schoolbook Multiplication vs. Number Theoretic Transform (NTT) Multiplication

In [1]:
from assignment_6b_parameters import *
from assignment_6b_functions import *
from assignment_6b_NTT import *

POLYMUL = 0

In [2]:
def mult_poly_vec(M1: matrix, M2: matrix) -> matrix:
    global POLYMUL
    POLYMUL = 0
    out = []
    A = M1.rows()
    B = M2.columns()[0]
    for f in range(k):
        t1 = []
        t2 = []
        for g in range(k):
            a = A[f][g].list() # 1a
            b = B[g].list() # 1b
            c = [0 for _ in range(n*2)] 

            for i in range(n):
                for j in range(n):
                    c[i + j] = (c[i + j] + a[i] * b[j]) % q # 1c
                    POLYMUL += 1 # DO NOT EDIT OR REMOVE THIS LINE
            t1.append(c)

        for i in range(k-1):
            for j in range(n*2):
                c = (t1[i][j] + t1[i+1][j]) % q
                t2.append(c)
        out.append([RQ(t2)])
    return matrix(out)

def mult_poly_vec_test():
    global NTTMUL
    for i in range(10):
        print(f'Running test {i+1}...')
        A = [[[None] for _ in range(0, k)] for _ in range(0, k)]
        s = [None for _ in range(0, k)]
        e = [None for _ in range(0, k)]

        # A is a k*k dimension matrix of polynomials with n terms
        for i in range(0, k):
            for j in range(0, k):
                A[i][j] = RandPolyUniform(n)
        A = matrix(A)

        for i in range(0, k):
            s[i] = [RandPolyCbd(n)]
        s = matrix(s)

        nttr, nttm = nttmul(A,s)
        assert(nttr == mult_poly_vec(A,s))
    print()
    print('All tests passed!')
    print()
    print('Total Multiplications for Schoolbook', POLYMUL)
    print('Total Multiplications for NTT', nttm)

In [3]:
mult_poly_vec_test()

Running test 1...
Running test 2...
Running test 3...
Running test 4...
Running test 5...
Running test 6...
Running test 7...
Running test 8...
Running test 9...
Running test 10...

All tests passed!

Total Multiplications for Schoolbook 262144
Total Multiplications for NTT 14336


## Part D

## How many multiplications are reported by schoolbook?
- Schoolbook Method Reports 262,144 total multiplications.
- NTT Method Reports 14,336 total Multiplications.

- What is the percent decrease from Schoolbook to NTT?

$$ \text{Percent Decrease } = \frac{\text{Final Value} - \text{Start Value}}{\text{Start Value}} \times 100 \% \Longrightarrow  \frac{262144 - 14336}{\text{262144}} \times 100 \% = 94.5313\% \text{ decrease} $$